python/cuspatial/cuspatial/core/binpreds/feature_contains.py

```diff
@@ -87,47 +99,35 @@ def _compute_polygon_polygon_contains(self, lhs, rhs, preprocessor_result):
         return result
 
-    def _test_interior(self, lhs, rhs):
-        # We only need to test linestrings that are length 2.
-        # Divide the linestring in half and test the point for containment
-        # in the polygon.
-        size_two = rhs.sizes == 2
-        if (size_two).any():
-            center_points = _linestrings_to_center_point(rhs[size_two])
-            size_two_results = _false_series(len(lhs))
-            size_two_results.iloc[rhs.index[size_two]] = (
-                _basic_contains_count(lhs, center_points) > 0
-            )
-            return size_two_results
-        else:
-            return _false_series(len(lhs))
-
     def _compute_polygon_linestring_contains(
         self, lhs, rhs, preprocessor_result
     ):
+        # Count the number of points in lhs that are properly contained by
+        # rhs
         contains = _basic_contains_count(lhs, rhs).reset_index(drop=True)
-        intersects = self._intersection_results_for_contains(lhs, rhs)
-
-        # If a linestring has intersection but not containment, we need to
-        # test if the linestring is in the interior of the polygon.
-        final_result = _false_series(len(lhs))
-        intersection_with_no_containment = (contains == 0) & (intersects != 0)
-        interior_tests = self._test_interior(
-            lhs[intersection_with_no_containment].reset_index(drop=True),
-            rhs[intersection_with_no_containment].reset_index(drop=True),
+
+        # Count the number of point intersections (line crossings) between
+        # lhs and rhs.
+        # Also count the number of perfectly overlapping linestring sections.
+        # Each linestring overlap counts as two point overlaps.
+        (
+            point_intersects_count,
+            linestring_intersects_count,
+        ) = self._intersection_results_for_contains_linestring(lhs, rhs)
+
+        # Subtract the length of the linestring intersections from the length
+        # of the rhs linestring, then test that the sum of contained points
+        # is equal to that adjusted rhs length.
+        rhs_sizes_less_line_intersection_size = (
+            rhs.sizes - linestring_intersects_count
         )
-        interior_tests.index = intersection_with_no_containment[
-            intersection_with_no_containment
-        ].index
-        # LineStrings that have intersection but no containment are set
-        # according to the `intersection_with_no_containment` mask.
-        final_result[intersection_with_no_containment] = interior_tests
-        # LineStrings that do not are contained if the sum of intersecting
-        # and containing points is greater than or equal to the number of
-        # points that make up the linestring.
-        final_result[~intersection_with_no_containment] = (
-            contains + intersects >= rhs.sizes
+        rhs_sizes_less_line_intersection_size[
+            rhs_sizes_less_line_intersection_size <= 0
+        ] = 1
```

Comment:
1. Why is it possible that there can be rows that has fewer vertices in the linestring to the number of intersection points?

2. Why do you need to explicitly set these rows to 1?

Reply:
Thanks for investigating this logic. This step is important to cover the tests of polygon.contains(linestring) where the LineString is contained only by the border off the polygon. In the above line, the number of points in the linestring is reduced by the number of points in the border-overlap region:

```
rhs_sizes_less_line_intersection_count = rhs.sizes - linestring_intersects_count
```
This is because LineString segments that overlap with the boundary of a polygon are not used in the counting of the vertices that may be contained in the polygon. Consider the following three examples:

![image](https://github.com/rapidsai/cuspatial/assets/410931/5fa9e534-eee9-458d-af47-a63c193f6911)

Note that each shape has examples of the LineString overlapping the boundary of the polygon. In the first case, an edge overlaps from the corner, then goes inside to an interior point, then returns to the edge, then overlaps to the corner. The length of the LineString is 5. The first overlapping segment subtracts 2 from the length of the linestring, and the second overlapping segment subtracts another 2, leaving a linestring of length 1. 1 point is contained, and therefor the linestring is contained.

In the second example the same pattern occurs except the middle point goes out of the polygon instead of in. The same points are subtracted from the length of the linestring, but the number of contained points is 0, less than the (remaining) length of the linestring of 1, so it is not contained.

Finally, the LineString only overlaps with the edge of the polygon, on the left and on the bottom. This returns a LineString from pairwise-linestring-intersection of length 3, which is subtracted from the length of the LineString, also 3. `rhs_sizes_less_line_intersection_size` is now 0 for this example LineString, so `contains + point_intersects_count == 0`, which would return `contains = True`. Therefore I set the minimum LineString length to 1, so that some LineString point must be contained in order for `.contains` to evaluate to true.

Happy to chat this through with you in person if you can think of a more parsimonious approach.

Reply:
> Note that each shape has examples of the LineString overlapping the boundary of the polygon. In the first case, an edge overlaps from the corner, then goes inside to an interior point, then returns to the edge, then overlaps to the corner. The length of the LineString is 5. The first overlapping segment subtracts 2 from the length of the linestring, and the second overlapping segment subtracts another 2, leaving a linestring of length 1. 1 point is contained, and therefor the linestring is contained.
> 
> In the second example the same pattern occurs except the middle point goes out of the polygon instead of in. The same points are subtracted from the length of the linestring, but the number of contained points is 0, less than the (remaining) length of the linestring of 1, so it is not contained.

Pardon me from not following the reasoning here. I took `linestring_intersects_count` meaning the number of vertices of the linestring that intersects with the ring of the polygon. Isn't for both cases, the value is 4? The subtracted result is 1 right?

> Finally, the LineString only overlaps with the edge of the polygon, on the left and on the bottom. This returns a LineString from pairwise-linestring-intersection of length 3, which is subtracted from the length of the LineString, also 3. rhs_sizes_less_line_intersection_size is now 0 for this example LineString, so contains + point_intersects_count == 0, which would return contains = True. Therefore I set the minimum LineString length to 1, so that some LineString point must be contained in order for .contains to evaluate to true.

This sounds a bit hacky to me. Also, at least the comparison should be `==` not `<=` right?